File: SkiKids/Utilities/DINCalculator.swift
import Foundation

struct DINCalculator {

    static func codeFromWeight(_ weightKg: Int) -> DINCode {
        for code in DINCode.allCases {
            if code.weightRange.contains(weightKg) {
                return code
            }
        }
        return .A
    }

    static func adjustCodeForHeight(_ code: DINCode, heightCm: Int) -> DINCode {
        var currentCode = code
        while heightCm > currentCode.heightThreshold {
            guard let next = currentCode.next else { break }
            currentCode = next
        }
        return currentCode
    }

    static func calculate(
        weightKg: Int,
        heightCm: Int,
        bslMm: Int,
        age: Int,
        ability: AbilityLevel
    ) -> DINResult {
        let baseCode = codeFromWeight(weightKg)
        let heightAdjustedCode = adjustCodeForHeight(baseCode, heightCm: heightCm)

        var warnings: [String] = []
        let adjustedCode: DINCode
        if age <= 12 && heightAdjustedCode != baseCode {
            adjustedCode = baseCode
            warnings.append("Height suggests a higher DIN code, but the lower (weight-based) setting was kept for child safety.")
        } else {
            adjustedCode = heightAdjustedCode
        }

        var dinValue = DINTable.lookup(code: adjustedCode, bsl: bslMm, ability: ability)

        var isJuniorAdjusted = false
        var lookupCode = adjustedCode

        if age <= 9 {
            if let previousCode = adjustedCode.previous {
                lookupCode = previousCode
                dinValue = DINTable.lookup(code: lookupCode, bsl: bslMm, ability: ability)
            }
            isJuniorAdjusted = true
        }

        dinValue = min(dinValue, 12.0)

        dinValue = (dinValue * 4).rounded() / 4

        let skierType: String
        switch ability {
        case .beginner: skierType = "Type I"
        case .intermediate: skierType = "Type II"
        case .advanced: skierType = "Type III"
        }

        if dinValue > 6.0 && age <= 12 {
            warnings.append("High DIN setting recommended. This should be verified by a certified ski technician. Some junior bindings do not support settings above 6.")
        }

        if weightKg < 10 {
            warnings.append("Very low weight detected. Verify minimum binding DIN capability with your ski technician.")
        }

        // Junior adjustment info is shown in the DIN detail view, not in the main warnings list

        return DINResult(
            value: dinValue,
            code: lookupCode.rawValue,
            skierType: skierType,
            isJuniorAdjusted: isJuniorAdjusted,
            warnings: warnings
        )
    }

    static func warnings(for child: (age: Int, heightCm: Int, weightKg: Int, ability: AbilityLevel, skiTypes: [SkiType])) -> [String] {
        var warnings: [String] = []

        if child.age <= 3 {
            warnings.append("For children under 3, consult a certified ski school. Sizing varies widely and individual assessment is recommended.")
        }

        if child.skiTypes.contains(.xcSkate) && child.age < 8 {
            warnings.append("Skate skiing is generally recommended for children 8 and older. Classic technique should be learned first.")
        }

        if child.ability == .advanced && child.age <= 6 {
            warnings.append("For children 6 and under, we recommend starting at Intermediate sizing even if ability is advanced, to maintain manoeuvrability.")
        }

        return warnings
    }
}

File: react-native/src/calculators/din-calculator.ts
import { AbilityLevel, DINCode, DINResult, SkiType } from "../models/types";
import { DIN_CODE_METADATA, DIN_CODE_ORDER, dinTableLookup } from "./din-table";

export function codeFromWeight(weightKg: number): DINCode {
  for (const code of DIN_CODE_ORDER) {
    const meta = DIN_CODE_METADATA[code];
    if (weightKg >= meta.weightMin && weightKg <= meta.weightMax) {
      return code;
    }
  }
  return DINCode.A;
}

export function adjustCodeForHeight(code: DINCode, heightCm: number): DINCode {
  let currentCode = code;
  while (heightCm > DIN_CODE_METADATA[currentCode].heightThreshold) {
    const idx = DIN_CODE_ORDER.indexOf(currentCode);
    if (idx + 1 >= DIN_CODE_ORDER.length) break;
    currentCode = DIN_CODE_ORDER[idx + 1];
  }
  return currentCode;
}

export function calculate(
  weightKg: number,
  heightCm: number,
  bslMm: number,
  age: number,
  ability: AbilityLevel
): DINResult {
  const baseCode = codeFromWeight(weightKg);
  const heightAdjustedCode = adjustCodeForHeight(baseCode, heightCm);

  const warnings: string[] = [];
  let adjustedCode: DINCode;

  if (age <= 12 && heightAdjustedCode !== baseCode) {
    adjustedCode = baseCode;
    warnings.push(
      "Height suggests a higher DIN code, but the lower (weight-based) setting was kept for child safety."
    );
  } else {
    adjustedCode = heightAdjustedCode;
  }

  let dinValue = dinTableLookup(adjustedCode, bslMm, ability);

  let isJuniorAdjusted = false;
  let lookupCode = adjustedCode;

  if (age <= 9) {
    const idx = DIN_CODE_ORDER.indexOf(adjustedCode);
    if (idx > 0) {
      lookupCode = DIN_CODE_ORDER[idx - 1];
      dinValue = dinTableLookup(lookupCode, bslMm, ability);
    }
    isJuniorAdjusted = true;
  }

  dinValue = Math.min(dinValue, 12.0);
  dinValue = Math.round(dinValue * 4) / 4;

  let skierType: string;
  switch (ability) {
    case AbilityLevel.Beginner:
      skierType = "Type I";
      break;
    case AbilityLevel.Intermediate:
      skierType = "Type II";
      break;
    case AbilityLevel.Advanced:
      skierType = "Type III";
      break;
    default:
      skierType = "Type I";
  }

  if (dinValue > 6.0 && age <= 12) {
    warnings.push(
      "High DIN setting recommended. This should be verified by a certified ski technician. Some junior bindings do not support settings above 6."
    );
  }

  if (weightKg < 10) {
    warnings.push(
      "Very low weight detected. Verify minimum binding DIN capability with your ski technician."
    );
  }

  return {
    value: dinValue,
    code: lookupCode,
    skierType,
    isJuniorAdjusted,
    warnings,
  };
}

export function getWarnings(child: {
  age: number;
  heightCm: number;
  weightKg: number;
  ability: AbilityLevel;
  skiTypes: SkiType[];
}): string[] {
  const warnings: string[] = [];

  if (child.age <= 3) {
    warnings.push(
      "For children under 3, consult a certified ski school. Sizing varies widely and individual assessment is recommended."
    );
  }

  if (child.skiTypes.includes(SkiType.XCSkate) && child.age < 8) {
    warnings.push(
      "Skate skiing is generally recommended for children 8 and older. Classic technique should be learned first."
    );
  }

  if (child.ability === AbilityLevel.Advanced && child.age <= 6) {
    warnings.push(
      "For children 6 and under, we recommend starting at Intermediate sizing even if ability is advanced, to maintain manoeuvrability."
    );
  }

  return warnings;
}
